### PR TITLE
Score a job GitHub never scheduled as dropped, not as a script error

### PR DESCRIPTION
**Pull request by groeneai**
Related: https://github.com/ClickHouse/ClickHouse/pull/116260 (witness, one abandoned job)
Related: https://github.com/ClickHouse/ClickHouse/pull/113737 (witness, eight of them, surviving two re-runs)
Related: https://github.com/ClickHouse/ClickHouse/pull/115745 (why the retry layer is untouched; no open issue exists for this class)

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
...

### Description

Nothing here retries, reschedules or filters any job. The change is only how praktika *scores* a job that produced no result.

GitHub reports an undocumented fifth `needs.<job>.result` value, `abandoned`, for a job it queued and never assigned a runner. `_finish_workflow` matched only `cancelled`/`canceled`, so such a job fell through to the `else` branch and was scored `ERROR` with `Job failed to produce Result due to a script error or CI runner issue` for a job that ran zero commands. `Mergeable Check` then said `Failed: <job name>`, naming a job that never started, and `is_cancelled` was never set. Once the node is `ERROR` it is `is_completed`, so a re-run's carried-forward `cancelled` record cannot repair it either.

I read the verdicts out of the `Finish Workflow` job logs, where the `toJson(needs)` heredoc is echoed: jobs 98272432828 and 98250718152. Matching every `needs` key against the jobs API over 4 attempts gives 688 job/verdict pairs with zero off-diagonal: `abandoned` pairs with zero-step-cancelled 10/10, `cancelled` 8/8. Report for one witness: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=116260&sha=9dd6a67f4ee2c7272ae707a47769a9a9efdcefbe&name_0=PR

The merge gate deliberately still blocks: `dropped_results` keeps `Mergeable Check` at FAIL and `DROPPED` still maps to a GitHub `error` status, so a check that did not run is not laundered into green. Only the description changes, from `Failed: <job name>` to `Failed: 0, Dropped: 1`.

Out of scope, and disclosed: the abandonment itself is GitHub-side with no in-repo lever, and job retry lives in the out-of-repo Lambda, per @ maxknv on #115745. The class is episodic rather than steady - it took 10 jobs across two of my PRs on 2026-08-26, while a 300-run sample of `pull_request.yml` runs started over the following 11 hours holds none.

Nothing committed pins this: https://github.com/ClickHouse/ClickHouse/pull/115650 removes `ci/tests` and the `CI Tests` job, so I validated with a throwaway harness per `AGENTS.md`.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=116648&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/116648](https://github.com/search?q=head%3Async-upstream%2Fpr%2F116648+org%3AClickHouse+type%3Apr&type=pullrequests)]

### Version info
- Merged into: `26.9.1.1110` (included in `26.9` and later)

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary><b>Internal second-model review: adjudication log</b> (click to expand)</summary>

Pre-publication review by an independent model (engine: codex; 3 findings). I fixed none of them
in code; the reasons are below, each with a citation rather than an argument.

| # | Sev | Finding | Verdict | Evidence / action |
|---|-----|---------|---------|---|
| 1 | ⚠️ | `abandoned` does not prove a job never started, so the new branch can score a runner loss as `DROPPED` (`native_jobs.py:1089-1103`) | DISAGREE | Premise granted, but the same branch already scores mid-run `cancelled` jobs `DROPPED` on master, so that is existing behaviour rather than something added here, and the note wording (`Job did not finish, GitHub reported [<verdict>]`) holds either way. Nothing turns green: `dropped_results` keeps the gate at `FAIL` (`native_jobs.py:1135`), `DROPPED` maps to a GitHub `error` (`gh.py:1269`), and it counts as failed in `Result._update_status`. The one call this drops, `add_workflow_error`, reaches only `check_report_messages.py`, which always posts green and never blocks; the reason moves to the node's `notes`, which the report renders. |
| 2 | ⚠️ | No automated test supplies an `abandoned` verdict, so reverting the fix would stay green (`ci/tests/test_result.py:273-353`) | AGREE on the fact, DISAGREE on the remedy | `.claude/CLAUDE.md:190` says a throwaway test for a CI script need not be committed, and #115650 is removing `ci/tests` entirely (updated 2026-08-27). I validated with an uncommitted 21-arm harness instead: 12 failed / 38 passed before the edit, 50 passed after, with the reddening arms built on the real captured 174-key `needs` payload. The fall-through log line now names the unrecognised verdict, so the next undocumented token identifies itself instead of being silent. If you would rather have a committed arm, say so and I will add one. |
| 3 | ⚠️ | `is_cancelled` stays latched after a re-run recovers the job (`native_jobs.py:1101`) | AGREE on the fact, DISAGREE on the severity | Confirmed, and pre-existing: that line latches today for every `cancelled` verdict, and a `DROPPED` node is never promoted back because `native_jobs.py:1066` short-circuits ahead of the classifier. Its only reader in the repository is the Slack emoji selector (`lambda_slack_worker.py:240`), so the visible effect is one flag emoji on a run the same feed still marks successful. The merge gate itself recovers: the job's own node goes `OK` and `dropped_results` is empty. Left alone to keep this change to the scoring only. |

Severity: ❌ blocker / ⚠️ major / 💡 nit. DISAGREE verdicts carry recorded evidence and are
terminal per finding.

</details>

Session id: cron:clickhouse-review-slot-9:20260827-050500

**Comment by groeneai** [comment]
<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|----------|--------|
| a | **Deterministic repro?** | Yes. The real `${{ toJson(needs) }}` payload captured from the `Finish Workflow` log of job 98272432828 (174 keys, 2 `abandoned`) is replayed through the unmodified `_finish_workflow`, which scores the job `ERROR` on every run. The base-tree output matches what CI actually published field for field: node `ERROR`, `ext.errors[0].message` string-equal to `ResultInfo.NOT_FINALIZED`, `start_time`/`duration` null, no files or links, no `is_cancelled`. Not a rate. |
| b | **Root cause explained?** | GitHub reports an undocumented fifth `needs.<job>.result` value, `abandoned`, for a job it queued and never assigned a runner. `native_jobs.py:1093` matched only `cancelled`/`canceled`, so the job took the `else` branch and was scored `ERROR` + `NOT_FINALIZED` ("script error") despite running zero commands, `is_cancelled` was never set, and `Mergeable Check` reported `Failed: <job name>`. The node is then `is_completed` (`result.py:213-214`), so a re-run's carried-forward `cancelled` record cannot reclassify it. |
| c | **Fix matches root cause?** | Yes. One missing value is added to an existing, intended whitelist (introduced by `15eddc5e424e238`, "handle cancelled workflow status"). No guard, no widened bound, no defensive check, and nothing in scheduling, digests, caches, `pipeline_status` or the retry layer is touched. |
| d | **Test intent preserved / new tests added?** | No test is added under `ci/tests/`: that directory is under a deletion directive (#115650 is open, `+9/-24715`, updated 2026-08-27), so a test there would be removed. Coverage is a 21-arm harness run before and after the change. The merged test that pins the deliberate absent-verdict `ERROR` path passes unchanged, and three matrix rows (`""`, `None`, job absent from `needs`) pin that same path independently. |
| e | **Both directions demonstrated?** | Yes. Base tree: 12 failed / 38 passed, including `assert 'ERROR' == 'DROPPED'` on both the minimal and the real-payload arms, the false `Failed: <job name>` on the merge-gate arm, and the fall-through log line lacking the verdict. Fixed tree: 50 passed. The whole `ci/tests/` suite was also run on both arms (what the `CI Tests` job runs): zero failures introduced, 12 pre-existing environmental failures identical on both sides. |
| f | **Fix is general across code paths?** | Re-measured with positive controls. `abandoned` occurs nowhere else in `ci/` or `.github/` (0 hits; control on the same grep: `cancelled` 519 hits). `native_jobs.py:1091` is the only place a `needs` verdict is read for classification; `_environment.py` only loads and returns the dict. The two `.github/` classifiers score the jobs/check-runs API `conclusion` instead (`failure`, `timed_out`, `action_required`, `startup_failure`); the measured conclusion for this class is `cancelled`, which none of them treat as a failure, so neither needs a change. |
| g | **Fix generalizes across inputs (params/datatypes/wrappers)?** | Verdict matrix, 14 rows: `abandoned` in three cases (the classifier lowercases, so the new value rides the same normalisation as the old ones), the two pre-existing spellings plus one uppercase, and the boundary and negative values `success`, `failure`, `skipped`, empty, `None`, an unknown verdict, a padded `" abandoned"` (deliberately not laundered), and a job with no `needs` entry at all. Type wrappers do not apply: the input is one string field. |
| h | **Backward compatible? (maintainer-approved exception only)** | Nothing to migrate: no setting, no serialization format, no `SettingsChangesHistory.cpp` (not a server change). One bounded behaviour change to an existing path: `cancelled`/`canceled` nodes now also carry the note, rendered `... [cancelled]`, which is why the wording is "did not finish" rather than "did not run" (`runner.py:157-178` publishes a `RUNNING` result before the job body, so such a node may genuinely have started). Pinned by its own arm. |
| i | **Invariants and contracts preserved?** | The invariant is that a job GitHub never ran is never scored as a failure of its own script, and never as green. Not greened: `DROPPED` maps to a GitHub `error` status (`gh.py:1269`) and `dropped_results` keeps the merge gate at `FAIL` (`native_jobs.py:1132-1140`), asserted in both directions. Persistence: the note is added before the existing `workflow_result.dump()`, and `notes` is in the ext key set the report assembly preserves, so it reaches the uploaded file. The branch's other post-conditions (`is_cancelled`, `dropped_results.append`, `update_final_report`, the `continue` that keeps the node out of `failed_results`) run unchanged and in the same order. |

</details>

Session id: cron:clickhouse-impl-slot-7:20260827-041000

**Comment by groeneai** [comment]
cc @leshikus, could you review this? It adds GitHub's undocumented `abandoned` verdict to the `_finish_workflow` classifier, so a job that GitHub queued and never assigned a runner is scored `DROPPED` with the verdict recorded, instead of `ERROR` with `Job failed to produce Result due to a script error or CI runner issue`. The merge gate still blocks on it.

File: ci/praktika/native_jobs.py [hunk 1]
```diff
@@ -1090,11 +1090,13 @@ def _finish_workflow(workflow, job_name):
             normalized_name = Utils.normalize_string(result.name)
             gh_job = workflow_job_data.get(normalized_name, {})
             gh_job_result = (gh_job.get("result") or "").lower()
-            if gh_job_result in ("cancelled", "canceled"):
+            # `abandoned` is GitHub's undocumented verdict for a job it queued and never assigned a runner.
+            if gh_job_result in ("cancelled", "canceled", "abandoned"):
```
[the remark this answers is not in the document]
Comment (groeneai) [reply ci/praktika/native_jobs.py:1094]: Confirmed by measurement, and the gap is total rather than partial: with `abandoned` removed from the tuple, every committed test that exercises `_finish_workflow` still passes (`test_result.py`, `test_pipeline_status_non_blocking.py`, `test_job_filtering.py`, 85 passed on both trees). `abandoned` occurs nowhere else under `ci/` or `.github/`, and the one source-inspecting test reads only `allow_failure`, so nothing committed pins this.

I wrote the arm you specified and ran it rather than declining on principle. It is load-bearing:

| tree | node `DROPPED` | carries the note | no `NOT_FINALIZED` |
|---|---|---|---|
| this branch | pass | pass | pass |
| `abandoned` removed from the tuple | fail | fail | fail |
| `add_note` line deleted | pass | fail | pass |

Two controls hold in all three trees: a `cancelled` verdict still becomes `DROPPED`, and an unrecognized verdict still reaches `ERROR` with `NOT_FINALIZED`. So the first mutant is narrow rather than a blanket break, and the note assertion is not carried by the status assertion.

I am not committing it, and the reason is the location rather than the arm. https://github.com/ClickHouse/ClickHouse/pull/115650 deletes `ci/tests/test_result.py`, `ci/jobs/ci_tests_job.py` and the `CI Tests` job definition; it is open and mergeable, and its predecessor https://github.com/ClickHouse/ClickHouse/pull/114675 merged at +0/-6543. `AGENTS.md` line 190: "When you want to test changes in the CI scripts, you can write and run a throwaway test, but you don't need to commit it to the repository, because we don't have automated CI tests." That is the form I used, and the description now records that the change is verified but unpinned.

**Review by qoega — APPROVED** [review]
